Allow User-Configured Input Spin State Orders From The GUI & Parameter File

cailafinn opened the pull request:
### Description of work

#### Summary of work
Adds two adjacent features to support the polarised reduction through the reflectometry interface. 
##### Wildes
- Allows the Flipper configuration (order of the spin states in the input workspace) to be given via the instrument parameter file. A default has been defined for POLREF. 
##### Fredrikze
- A rarer polarisation correction that would have different spin state orders on an experiment-to-experiment basis. 
- To facilitate this a new field has been added to the Experiment Settings tab of the ISIS Reflectometry Interface. 
- This field is then passed to the Fredrikze polarisation correction algorithm. 
- The reduction algorithm throws an error if the reduction expected to perform a Wildes correction but a value was entered for the Fredrikze order. 

Fixes #37617 

### To test:
1. Open the ISIS Reflectometry Interface
2. Switch the Instrument to POLREF
3. Enter a run of 37125 and an angle of 2.3 on the runs tab.
4. Run this script:
```python
eWs = CreateSampleWorkspace('Histogram', Function='User Defined', UserDefinedFunction='name=UserFunction,Formula=(1 + tanh(0.0733 * 12 * x * 0.2))/2',XUnit='Wavelength', xMin='1',XMax='8', BinWidth='1', NumBanks='1', BankPixelWidth='1')
# Wildes  {"P1", "P2", "F1", "F2"}
eff_wildes = JoinISISPolarizationEfficiencies(P1=eWs, P2=eWs, F1=eWs, F2=eWs )

# Fredrikze {"Pp", "Ap", "Rho", "Alpha"}
eff_fredrikze = JoinISISPolarizationEfficiencies(Pp=eWs, Ap=eWs, Rho=eWs, Alpha=eWs )
```
5. Swap to the experiment settings tab
6. Set the Polarization Corrections to workspace.

#### Wildes
1. Select the `eff_wildes` workspace from the Polarization Efficiencies dropdown
2. Run the reduction on the Runs tab. Given this run is a 2 workspace group, and the setting in the parameter file is for 4 workspaces, we expect this to fail. Change the WildesFlipperConfig in the POLREF_Parameter file from `00,01,10,11` to `0,1` to see it succeed. 

#### Fredrikze
1. Select the `eff_fredrikze` workspace from the polarization efficiencies dropdown
2. Enter a fredrikze spin state order into the text entry box and run the reduction. 
3. It should fail when four entries are given E.g `pa,ap,pp,aa`
4. It should succeed when two entries are given `p, a`

---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

cailafinn committed:
Return spin states from parameter file for Wildes

RE #37617

cailafinn committed:
Add fredrikze spin state line-edit to gui

RE #37617

cailafinn committed:
Add getters and setters to view

RE #37617

cailafinn committed:
Implement input disabling logic

RE #37617

cailafinn committed:
Add spin state order to Pol Corr model

RE #37617

cailafinn committed:
Update model from presenter

RE #37617

cailafinn committed:
Validate Fredrikze spinstate order at alg level

RE #37617

cailafinn committed:
Add documentation

RE #37617

cailafinn committed:
Include new field in encoder and decoder

This is needed for batch saving/loading and for project recovery.

RE #37617

adriazalvarez reviewed (changes requested):
I have followed the testing instructions, I found an issue with the validation of the flipper setup in the idf file that maybe we should look into in a separate issue. 
Aside that, the code changes look good and the GUI is behaving as expected. Thanks for updating the testing instructions too. 
I have added some other minor comments.

cailafinn commented on Framework/Reflectometry/test/ReflectometryReductionOneAuto3Test.h:
```diff
@@ -1310,6 +1310,105 @@ class ReflectometryReductionOneAuto3Test : public CxxTest::TestSuite {
+    // We're setting this to an invalid value to catch it on purpose later.
+    input_group[0]->instrumentParameters().addString(input_group[0]->getInstrument()->getComponentID(),
+                                                     "WildesFlipperConfig", "01,01,10");
+    auto efficiencies = createPolarizationEfficienciesWorkspace("Wildes");
```
I had a bit of a look into it, but I think it's probably better off being a separate issue. I've created this one here: 
- #38837

cailafinn committed:
Use std::optional where appropriate

RE #37617

cailafinn commented on Framework/Reflectometry/src/ReflectometryReductionOneAuto3.cpp:
```diff
@@ -56,8 +56,8 @@ void validate(const std::string &method) {
 static const std::string PA{"PA"};
-static const std::string FLIPPERS_NO_ANALYSER{"0, 1"};
-static const std::string FLIPPERS_FULL{"00, 01, 10, 11"};
+static const std::string DEFAULT_FLIPPERS_NO_ANALYSER{"0, 1"};
```
Ah yes, I see what you mean. I think it's probably worth switching to a SpinStateValidator for the algorithm in this PR to be honest. I'll implement that.

cailafinn committed:
Use a  SpinStateValidator for passing to Wildes

Uses a SpinStateValidator to match the validator in
PolarizationCorrectionWildes. This additionally fixes a problem where
not including spaces when providing spin states would result in an
error.

RE #37617

adriazalvarez reviewed (approved):
Thanks for the changes and for adding the validator.
